pallets/pools/src/impls.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use common_traits::{CurrencyPair, PriceValue};
+use common_types::CurrencyId;
+
+use super::*;
+
+impl<T: Config> PoolInspect<T::AccountId, T::CurrencyId> for Pallet<T> {
+	type PoolId = T::PoolId;
+	type TrancheId = T::TrancheId;
+	type Rate = T::BalanceRatio;
+	type Moment = Moment;
+
+	fn pool_exists(pool_id: Self::PoolId) -> bool {
+		Pool::<T>::contains_key(pool_id)
+	}
+
+	fn tranche_exists(pool_id: Self::PoolId, tranche_id: Self::TrancheId) -> bool {
+		Pool::<T>::get(pool_id)
+			.and_then(|pool| pool.tranches.tranche_index(&TrancheLoc::Id(tranche_id)))
+			.is_some()
+	}
+
+	fn get_tranche_token_price(
+		pool_id: Self::PoolId,
+		tranche_id: Self::TrancheId,
+	) -> Option<PriceValue<T::CurrencyId, T::BalanceRatio, Moment>> {
+		let now = Self::now();
+		let mut pool = Pool::<T>::get(pool_id)?;
+
+		// Get cached nav as calculating current nav would be too computationally expensive
+		if let Some((nav, nav_last_updated)) = T::NAV::nav(pool_id) {
```

Comment:
To simplify this block, I suggest:

1. Instead of having a `if <> { huge_block } else { None }`, do:
``` rust
let (nav, nav_last_updated) = T::NAV::nav(pool_id)?;
```

and thus flatten this whole block